### PR TITLE
Shipping panel normalization

### LINKED ISSUE #1404: Shipping Dashboard Panels
As we're preparing to add additional shipping providers, we need to make some UIX updates to the Shipping dashboard panels, here's a few...
- [ ] Open / Close collapse of provider panel. 
- [ ] Sort Disabled to bottom so that custom/active providers are first
- [ ] Use a priority to determine order of enabled providers

Use case:
I have three providers:
- Local Home Delivery
- Shippo
- Flat Rates

Where "LHD" is a custom plugin, and Shippo is enabled, but I don't want to see it's panel contents, not my disabled Flat Rates. "LHD" is the panel that my shop uses for tracking shipping.

Each of these should also respect viewing permissions from the user's package/role.

### PR DESCRIPTION
**Pull request by aaronjudd**
Resolves #1404
WIP #421

Moves more i18n into plugins/core/shipping
Moves flat rates into plugins/included/shipping-rates
Normalized settings panels (to look like other panels)
Refactored shipping UI to griddle with editing.
Adds editable Shippo carrier grid.
Adds `onGetShippingRates` Event hook, moves processing to modules
Normalized checkout template, removed template.inheritance

TODO -> introduce summary grouped rates (new issue)

### CODE REVIEW

**Comment by zenweasel** [comment]
Changing the label for a shipper does not seem to be reflected in checkout

**Comment by zenweasel** [comment]
The Enabled/Disabled state is not being properly reflected for each of the Flat Rate methods (i.e. if you enable one, each one will show enabled and vice versa)

**Comment by zenweasel** [comment]
Is there a way to disable/remove a Flat Rate method once you add it?

File: imports/plugins/core/shipping/client/templates/checkout/shipping.js [hunk 1]
```diff
@@ -1,37 +1,48 @@
...
+// returns multiple methods
+function cartShippingQuotes(currentCart) {
   const cart = currentCart || Cart.findOne();
+  const shipmentQuotes = [];
+
   if (cart) {
     if (cart.shipping) {
-      if (cart.shipping[0].shipmentQuotes) {
-        return cart.shipping[0].shipmentQuotes;
+      for (shipping of cart.shipping) {
```
Comment (zenweasel) [inline imports/plugins/core/shipping/client/templates/checkout/shipping.js]: Should this be `const shipping`?

File: imports/plugins/core/shipping/client/templates/checkout/shipping.js [hunk 2]
```diff
@@ -1,37 +1,48 @@
...
+        if (shipping.shipmentQuotes) {
+          for (quote of shipping.shipmentQuotes) {
```
Comment (zenweasel) [inline imports/plugins/core/shipping/client/templates/checkout/shipping.js]: Here as well?

File: imports/plugins/core/taxes/server/methods/methods.js [hunk 3]
```diff
@@ -97,7 +97,7 @@ export const methods = {
     if (pkg && pkg.enabled === true && pkg.settings.rates.enabled === true) {
       Logger.debug("Calculating custom tax rates");
 
-      if (typeof cartToCalc.shipping !== "undefined") {
+      if (typeof cartToCalc.shipping !== "undefined" && cartToCalc.items !== "undefined") {
```
Comment (zenweasel) [inline imports/plugins/core/taxes/server/methods/methods.js]: Should this be `typeof cartToCalc.items`?

File: imports/plugins/included/shippo/client/settings/carriers.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        displayName: i18next.t(`admin.shippingGrid.${field}`)
+      };
+      customColumnMetadata.push(columnMeta);
+    });
+
+    // filter and extract shipping methods
+    // from flat rate shipping provider
+    function transform(results) {
+      const result = [];
+      for (method of results) {
```
Comment (zenweasel) [inline imports/plugins/included/shippo/client/settings/carriers.js]: `const method`

File: imports/plugins/included/shippo/server/methods/shippo.js [hunk 5]
```diff
@@ -1,9 +1,13 @@
 /* eslint camelcase: 0 */
+import { Meteor } from "meteor/meteor";
+import { check } from "meteor/check";
 import { Reaction } from "/server/api";
 import { Packages, Accounts, Shops, Shipping, Cart, Orders } from "/lib/collections";
 import { ShippoPackageConfig } from "../../lib/collections/schemas";
 import { ShippoApi } from "./shippoapi";
 
+const shippingRoles = ["admin", "owner", "shipping", "reaction-shippo"];
```
Comment (zenweasel) [inline imports/plugins/included/shippo/server/methods/shippo.js]: Are we defining this in more than one place?

File: server/methods/core/orders.js [hunk 6]
```diff
@@ -843,11 +843,9 @@ export const methods = {
             });
 
             // Temporarily(?) put here the Shippo's method/label purchasing.After a succesfull capture fund
-            if (order.shipping[0].shipmentMethod.shippoMethod) {
+            if (order.shipping[0].shipmentMethod.settings) {
```
Comment (prosf) [inline server/methods/core/orders.js]: A hook implementation( "eg onPaymentCapture")  maybe could be a nice improvement

**Comment by prosf** [comment]
Nice work!

File: imports/plugins/included/shipping-rates/client/templates/settings/rates.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            isEditing: false,
+            editingId: null
+          });
+        }
+      }
+    });
+  },
+  "click .shipping-grid-row": function (event) {
+    // toggle all rows off, then add our active row
+    $(".shipping-grid-row").removeClass("active");
```
Comment (aaronjudd) [inline imports/plugins/included/shipping-rates/client/templates/settings/rates.js]: Good point, I'll capture these two suggestions @prosf's hook in a quick update.. there's a few more thing cleanup items.

File: imports/plugins/core/ui-grid/client/griddle.js [hunk 8]
```diff
@@ -73,7 +74,11 @@ const MeteorGriddle = React.createClass({
       }, options));
     }
 
-    const results = this.props.collection.find(this.state.query, options).fetch();
+    // optional transform of collection for grid results
+    let results = this.props.collection.find(this.state.query, options).fetch();
+    if (this.props.transform) {
```
Comment (aaronjudd) [inline imports/plugins/core/ui-grid/client/griddle.js]: Good catch.

File: server/methods/core/orders.js (hunk 6 above)
Comment (aaronjudd) [inline server/methods/core/orders.js]: 👍  agreed. adding now.

**Pull request merged**